Backends Ed25519 implementation is not compatible with upstream

### Microsoft build of Go version

go 1.25

### What is your operating system and platform?

TestSignatureSelection is failing because struct scheme of our backends not fully compatible with upstream Ed25519 Private Key. This needs further investigation.

### Output of `go env` in your module/workspace:

```shell
go version go1.25-devel_4a6cf6eac98 Mon Jun 30 14:12:19 2025 +0100 darwin/arm64
mert@Mac src % ../bin/go env    
AR='ar'
CC='clang'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='clang++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/mert/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/mert/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/yl/_kmd0h517bj5k39lj_33bz2c0000gn/T/go-build2422851309=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/Users/mert/go/src/github.com/microsoft/go/go/src/go.mod'
GOMODCACHE='/Users/mert/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/mert/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/Users/mert/go/src/github.com/microsoft/go/go'
GOSUMDB='sum.golang.org'
GOTELEMETRY='on'
GOTELEMETRYDIR='/Users/mert/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/Users/mert/go/src/github.com/microsoft/go/go/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.25-devel_4a6cf6eac98 Mon Jun 30 14:12:19 2025 +0100'
GOWORK=''
PKG_CONFIG='pkg-config'
```

### What did you do?

I run TestSignatureSelection test.

### What did you see happen?

It failed with following logs:
--- FAIL: TestSignatureSelection (0.00s)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:67: skipping test[1] - not compatible with TLS FIPS mode
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:67: skipping test[8] - not compatible with TLS FIPS mode
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:75: test[11]: unexpected selectSignatureScheme error: tls: internal error: unsupported key (ed25519.PrivateKey)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:78: test[11]: expected signature scheme Ed25519, got SignatureScheme(0)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:82: test[11]: unexpected typeAndHashFromSignatureScheme error: unsupported signature algorithm: SignatureScheme(0)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:85: test[11]: expected signature algorithm 0xe4, got 0x0
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:75: test[12]: unexpected selectSignatureScheme error: tls: internal error: unsupported key (ed25519.PrivateKey)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:78: test[12]: expected signature scheme Ed25519, got SignatureScheme(0)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:82: test[12]: unexpected typeAndHashFromSignatureScheme error: unsupported signature algorithm: SignatureScheme(0)
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:85: test[12]: expected signature algorithm 0xe4, got 0x0
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:67: skipping test[13] - not compatible with TLS FIPS mode
    /Users/mert/go/src/github.com/microsoft/go/go/src/crypto/tls/auth_test.go:67: skipping test[14] - not compatible with TLS FIPS mode

### What did you expect to see?

Fully passing tests.

--- a/patches/0004-Use-crypto-backends.patch
+++ b/patches/0004-Use-crypto-backends.patch
@@ -67,7 +67,7 @@ Subject: [PATCH] Use crypto backends
  src/crypto/sha256/sha256_test.go              |  20 +-
  src/crypto/sha512/sha512.go                   |   2 +-
  src/crypto/sha512/sha512_test.go              |  24 ++-
- src/crypto/tls/auth_test.go                   |   8 +
+ src/crypto/tls/auth_test.go                   |   7 +
  src/crypto/tls/cipher_suites.go               |   2 +-
  src/crypto/tls/defaults_boring.go             |   2 +-
  src/crypto/tls/defaults_fips140.go            |   2 +-
@@ -90,7 +90,7 @@ Subject: [PATCH] Use crypto backends
  src/net/lookup_test.go                        |   3 +
  src/os/exec/exec_test.go                      |   9 +
  src/runtime/pprof/vminfo_darwin_test.go       |   6 +
- 86 files changed, 1161 insertions(+), 96 deletions(-)
+ 86 files changed, 1160 insertions(+), 96 deletions(-)
  create mode 100644 src/crypto/dsa/boring.go
  create mode 100644 src/crypto/dsa/notboring.go
  create mode 100644 src/crypto/ecdsa/badlinkname.go
@@ -2318,18 +2318,10 @@ index 080bf694f03652..8962a54f1bcddd 100644
  	})
  }
 diff --git a/src/crypto/tls/auth_test.go b/src/crypto/tls/auth_test.go
-index 727bdc4df76393..1a95932f002a71 100644
+index 727bdc4df76393..73dee851e1bb4a 100644
 --- a/src/crypto/tls/auth_test.go
 +++ b/src/crypto/tls/auth_test.go
-@@ -12,6 +12,7 @@ import (
- )
- 
- func TestSignatureSelection(t *testing.T) {
-+	t.Skip("Skipping this test until #1715 is resolved.")
- 	rsaCert := &Certificate{
- 		Certificate: [][]byte{testRSACertificate},
- 		PrivateKey:  testRSAPrivateKey,
-@@ -67,6 +68,13 @@ func TestSignatureSelection(t *testing.T) {
+@@ -67,6 +67,13 @@ func TestSignatureSelection(t *testing.T) {
  			t.Logf("skipping test[%d] - not compatible with TLS FIPS mode", testNo)
  			continue
  		}

--- a/patches/0004-Use-crypto-backends.patch
+++ b/patches/0004-Use-crypto-backends.patch
@@ -67,7 +67,7 @@ Subject: [PATCH] Use crypto backends
  src/crypto/sha256/sha256_test.go              |  20 +-
  src/crypto/sha512/sha512.go                   |   2 +-
  src/crypto/sha512/sha512_test.go              |  24 ++-
- src/crypto/tls/auth_test.go                   |   1 +
+ src/crypto/tls/auth_test.go                   |   8 +
  src/crypto/tls/cipher_suites.go               |   2 +-
  src/crypto/tls/defaults_boring.go             |   2 +-
  src/crypto/tls/defaults_fips140.go            |   2 +-
@@ -90,7 +90,7 @@ Subject: [PATCH] Use crypto backends
  src/net/lookup_test.go                        |   3 +
  src/os/exec/exec_test.go                      |   9 +
  src/runtime/pprof/vminfo_darwin_test.go       |   6 +
- 86 files changed, 1154 insertions(+), 96 deletions(-)
+ 86 files changed, 1161 insertions(+), 96 deletions(-)
  create mode 100644 src/crypto/dsa/boring.go
  create mode 100644 src/crypto/dsa/notboring.go
  create mode 100644 src/crypto/ecdsa/badlinkname.go
@@ -116,7 +116,7 @@ index f0e3575637c62a..9eab3b4e66e60b 100644
  package main
  
 diff --git a/src/cmd/dist/test.go b/src/cmd/dist/test.go
-index c2adb959460ac2..ee766afaca06e7 100644
+index c3f3a8324c507c..aedd8cd2902f27 100644
 --- a/src/cmd/dist/test.go
 +++ b/src/cmd/dist/test.go
 @@ -719,7 +719,7 @@ func (t *tester) registerTests() {
@@ -128,7 +128,7 @@ index c2adb959460ac2..ee766afaca06e7 100644
  		// Test standard crypto packages with fips140=on.
  		t.registerTest("GOFIPS140=latest go test crypto/...", &goTest{
  			variant: "gofips140",
-@@ -1199,6 +1199,11 @@ func (t *tester) internalLink() bool {
+@@ -1211,6 +1211,11 @@ func (t *tester) internalLink() bool {
  	if goos == "windows" && goarch == "arm64" {
  		return false
  	}
@@ -140,7 +140,7 @@ index c2adb959460ac2..ee766afaca06e7 100644
  	// Internally linking cgo is incomplete on some architectures.
  	// https://golang.org/issue/10373
  	// https://golang.org/issue/14449
-@@ -1362,12 +1367,11 @@ func (t *tester) registerCgoTests(heading string) {
+@@ -1374,12 +1379,11 @@ func (t *tester) registerCgoTests(heading string) {
  			// a C linker warning on Linux.
  			// in function `bio_ip_and_port_to_socket_and_addr':
  			// warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
@@ -2318,7 +2318,7 @@ index 080bf694f03652..8962a54f1bcddd 100644
  	})
  }
 diff --git a/src/crypto/tls/auth_test.go b/src/crypto/tls/auth_test.go
-index 727bdc4df76393..8fd9fc75cf05ef 100644
+index 727bdc4df76393..1a95932f002a71 100644
 --- a/src/crypto/tls/auth_test.go
 +++ b/src/crypto/tls/auth_test.go
 @@ -12,6 +12,7 @@ import (
@@ -2329,6 +2329,20 @@ index 727bdc4df76393..8fd9fc75cf05ef 100644
  	rsaCert := &Certificate{
  		Certificate: [][]byte{testRSACertificate},
  		PrivateKey:  testRSAPrivateKey,
+@@ -67,6 +68,13 @@ func TestSignatureSelection(t *testing.T) {
+ 			t.Logf("skipping test[%d] - not compatible with TLS FIPS mode", testNo)
+ 			continue
+ 		}
++
++		if fips140tls.Required() && test.expectedSigAlg == Ed25519 {
++			// CNG and SymCrypt do not support Ed25519.
++			t.Logf("skipping test[%d] - Ed25519 is not compatible with Microsoft TLS FIPS mode", testNo)
++			continue
++		}
++
+ 		savedGODEBUG := os.Getenv("GODEBUG")
+ 		os.Setenv("GODEBUG", savedGODEBUG+","+test.godebug)
+ 
 diff --git a/src/crypto/tls/cipher_suites.go b/src/crypto/tls/cipher_suites.go
 index 2a96fa69036a07..f3c19c7faf1c21 100644
 --- a/src/crypto/tls/cipher_suites.go
@@ -2831,7 +2845,7 @@ index e7369542a73270..ff52175e4ac636 100644
  	}
  }
 diff --git a/src/go/build/deps_test.go b/src/go/build/deps_test.go
-index eddaf170772e19..dbd96efb9a7dce 100644
+index 1997d90f2f20ba..f69862c628eea4 100644
 --- a/src/go/build/deps_test.go
 +++ b/src/go/build/deps_test.go
 @@ -517,7 +517,7 @@ var depsRules = `

